src/server.cpp

```diff
@@ -1100,6 +1103,50 @@ void Server::Receive(float timeout)
 	}
 }
 
+void Server::yieldToOtherThreads(float dtime)
+{
+	/*
+	 * Problem: the server thread and emerge thread compete for the envlock.
+	 * While the emerge thread needs it just once or twice for every processed item
+	 * the server thread uses it much more generously.
+	 * This is usually not a problem as the server sleeps between steps, which leaves
+	 * enough chance. But if the server is overloaded it's busy all the time and
+	 * - even with a fair envlock - the emerge thread can't get up to speed.
+	 * This generally has a much worse impact on gameplay than server lag itself
+	 * ever would.
+	 *
+	 * Workaround: If we detect that the server is overloaded, introduce some careful
+	 * artificial sleeps to leave the emerge thread(s) enough chance to do their job.
+	 *
+	 * In the future the emerge code could be reworked to exclusively use a result
+	 * queue, thereby avoiding this problem.
+	 */
+
+	if (dtime < getStepSettings().steplen)
+		return;
```

Comment:
I get the intent to avoid enabling this due to random fluctuations, but the suggestion is wrong.
Lock contention starts happening as soon as there is no sleep time in the ServerThread, this is precisely when dtime > steplen.

> Also, this check should be in ServerThread::run

I'll move it there

> because there we have more knowledge about how long things actually took.

however isn't it subtly wrong when I don't include the time spent on `Receive`?

Reply:
The emerge thread may even be already suffering if the dtime is almost the target dtime.

You could measure in `Server::Receive()` how long it has been waiting for packets (and maybe use `TryReceive` before as long as possible). And if the sleep time is not long enough, do the `yieldToOtherThreads`. Shouldn't be hard.

Reply:
I could also use `EnvAutoLock` to exactly measure the locking time, sure.
But is it worth to invest so much time into a (hopefully) temporary workaround?

Reply:
If you think it's too much work, and if you say it's just temporary, then I'm fine with this.

Reply:
I mean I currently have no plans to continue the work and convert emerging to a queue, but it's not that hard so I'm hoping:
1. that the current workaround is good enough with small enough downsides
2. someone else picks up that task soon *or* I change my opinion in 2 months